indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java

```diff
@@ -580,6 +606,12 @@ protected void reduce(
             persist(index, interval, finalFile, progressIndicator);
             toMerge.add(finalFile);
           }
+          persistExecutor.shutdown();
+          boolean loop = true;
+          do {
+            loop = !persistExecutor.awaitTermination(1, TimeUnit.SECONDS);
+            checkPersistException();
```

Comment:
Most places in code would use Futures, and then call `Futures.allAsList(futures).get(1, TimeUnit.HOURS)` or similar.

The catch with that approach is you need to be able to have the incremental index garbage collected, so you have to eliminate hard references to the incremental index in the future.